File: go/stack/22_Generate_Parentheses/22_Generate_Parentheses.go
package generateparentheses

func generateParenthesis(n int) []string {
	var result []string
	backTrack(0, 0, n, &result, "")
	return result
}

func backTrack(open int, closed int, n int, result *[]string, cur string) {
	if (open == closed) && (closed == n) && (open == n) {
		*result = append(*result, cur)
		return
	}
	if closed > open {
		return
	}

	if open < n {
		backTrack(open+1, closed, n, result, cur+"(")
	}
	if closed < n {
		backTrack(open, closed+1, n, result, cur+")")
	}
}

File: csharp/22-Generate-Parentheses.cs
public class Solution {
    public IList<string> GenerateParenthesis(int n) {
        var result = new List<string>();
        var seq = new StringBuilder();
        
        void backtrack(int open, int close) {
            if(seq.Length == n * 2) {
                result.Add(seq.ToString());
                return;
            } 
            
            if(open < n) {
                seq.Append("(");
                backtrack(open + 1, close);
                seq.Remove(seq.Length - 1, 1);
            }
            if(close < open) {
                seq.Append(")");
                backtrack(open, close + 1);
                seq.Remove(seq.Length - 1, 1);
            }
            
        }

        backtrack(0, 0);

        return result;
    }
}